experimentation/export_model_predictions.py

```diff
@@ -0,0 +1,148 @@
+}
+
+
+def get_disease_nodes(
+    take: int | None = None,
+    nxo: NXOntology[str] | None = None,
+) -> Iterable[str]:
+    """
+    If all roots are in the set of non-disease therapeutic areas (aka roots in EFO OTAR Slim),
+        then its non-disease
+    """
+    nxo = nxo or get_efo_otar_slim()
+    for n in islice(sorted(nxo.graph), take):
+        if any(
+            root not in NON_DISEASE_THERAPEUTIC_AREAS for root in nxo.node_info(n).roots
+        ):
+            yield n
+
+
+rs_cls_to_precision = {
+    "01-disease-subtype": "high",
+    "02-disease-root": "medium",
+    "03-disease-area": "low",
+}
+
+
+@dataclass
+class NodeLabelOutput:
+    """
+    Main output of the model inference tasks
+    """
+
+    identifier: str
+    precision: str
+    rs_classification: str | None
+    probas: list[float]
+
+
+def export_model_predictions(
+    precisions_file: Path = ROOT_DIR / "data/efo_otar_slim_v3.57.0_precisions.tsv",
+    features_file: Path = ROOT_DIR / "data/efo_otar_slim_v3.57.0_features.tsv",
+    take: int | None = None,
+    text_embeddings_enabled: bool = True,
+) -> None:
+    """
+    1. Train a model (the best performing one) on the entire training set
+    2. Run inference on the entire EFO graph (except the non-disease nodes)
+    3. Export both labels and the feature values for each node
+    """
+    assert not precisions_file.exists(), f"{precisions_file} already exists, aborting"
+    assert not features_file.exists(), f"{features_file} already exists, aborting"
+
+    # 1. Train model
+    X, y = read_training_data(filter_out_non_disease=True, take=take)
+
+    feature_pipeline: Pipeline = make_pipeline(
+        PrepareNodeFeatures(),
+        NodeInfoFeatures(),
+        NodeXrefFeatures(),
+        TextEmbeddingsTransformer.from_config(
+            enabled=text_embeddings_enabled,
+            pca_components=64,
+            use_lda=False,
+        ),
+        CatBoostDataFormatter(),
+    )
+    X_transform = feature_pipeline.fit_transform(X, y)
+    model = CatBoostClassifier(
+        eval_metric=BiasedMaeMetric(),
+        custom_metric=["MultiClass", "AUC", "F1"],
+        learning_rate=0.5,
+        iterations=5000,
+        metric_period=250,
+    )
+    model.fit(
+        X=Pool(
+            data=X_transform,
+            label=list(y),
+        )
+    )
+
+    # 2. Do inference on new version of the ontology
+    target_nodes: list[str] = list(get_disease_nodes(take=take))
+    target_features = feature_pipeline.transform(target_nodes)
+    target_labels = model.predict(target_features)
+    target_probas = model.predict_proba(target_features)
+    training_set_labels = {id_: str(label) for id_, label in zip(X, y, strict=True)}
+
+    # 3. Write labels and features into files
+    labels_output: list[NodeLabelOutput] = [
+        NodeLabelOutput(
+            identifier=node_id,
+            precision=rs_cls_to_precision[label[0]],
+            probas=probas.tolist(),
```

Comment:
The output probabilities should be in separate columns with a column name that makes it easy to know which applies to which class.

Reply:
okay